spec/cps.nim

```diff
@@ -0,0 +1,150 @@
+## Implements some macros and types to make writing code in continuation-
+## passing style easier.
+
+import std/macros
+
+{.experimental: "callOperator".}
+
+type
+  CellBase = ref object of RootObj
+    ## Header for a heap cell.
+    copy: (proc(x: CellBase): CellBase {.nimcall, raises: [].})
+  Cell[T] = ref object of CellBase
+    val: T
+
+  CellRef = object
+    ## Type-erased managed unique pointer to a heap cell. Needed for storing
+    ## the environment for continuations.
+    p: CellBase
+
+# --- cell implementation
+
+proc `=copy`(dst: var CellRef, src: CellRef) =
+  if dst.p != src.p: # do nothing for self assignments
+    if src.p.isNil:
+      dst.p = nil
+    else:
+      dst.p = src.p.copy(src.p)
+
+proc take[T](c: sink CellRef): T {.inline.} =
```

Comment:
```suggestion
proc take[T](c: sink CellRef): T {.inline.} =
  ## returns the value held by the cell, `c`, as type `T`
```

I was explaining the code to myself and I figured I'd write it down here

Reply:
I've used a slightly adjusted version of your suggestions, putting emphasis on the value being required to have the correct type `T`.